File: backend/src/modules/media/media.service.ts
import path from 'path';
import fs from 'fs/promises';
import { Media } from '@database/index';
import { AppError } from '@utils/AppError';
import { parsePagination } from '@utils/pagination';
import { env } from '@config/env';
import { ensureUploadDir } from './media.upload';
import { writeAuditLog } from '@utils/audit';
import { Op, WhereOptions } from 'sequelize';

const PUBLIC_BASE = '/uploads';

export async function listMedia(actorId: string, filters: {
  page?: string;
  limit?: string;
  search?: string;
  mimeType?: string;
  sort?: string;
  order?: 'asc' | 'desc';
}) {
  const { page, limit, offset, sort, order } = parsePagination(filters);

  const where: WhereOptions<Media> = {};
  if (filters.search) {
    where[Op.or as never] = [{ originalName: { [Op.like]: `%${filters.search}%` } }];
  }
  if (filters.mimeType) {
    where.mimeType = { [Op.like]: `${filters.mimeType}%` };
  }

  const { rows, count } = await Media.findAndCountAll({
    where,
    limit,
    offset,
    order: [[sort, order]],
    attributes: { exclude: ['deletedAt'] },
    distinct: true,
  });

  return {
    media: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function getMediaById(id: number) {
  const media = await Media.findByPk(id, { attributes: { exclude: ['deletedAt'] } });
  if (!media) {
    throw AppError.notFound('Media not found');
  }
  return media;
}

export async function saveUpload(actorId: string, file: Express.Multer.File, altText?: string) {
  if (!file) {
    throw AppError.badRequest('No file uploaded');
  }

  const ext = path.extname(file.originalname).toLowerCase();
  const baseDir = ensureUploadDir();
  const url = `${PUBLIC_BASE}/${file.filename}`;

  const media = await Media.create({
    userId: Number(actorId),
    fileName: file.filename,
    originalName: file.originalname,
    mimeType: file.mimetype,
    fileSize: file.size,
    storagePath: path.relative(process.cwd(), path.join(baseDir, file.filename)),
    url,
    altText: altText ?? null,
  });

  await writeAuditLog({
    actorId: Number(actorId),
    action: 'MEDIA_UPLOADED',
    module: 'media',
    entityType: 'media',
    entityId: media.id,
    newValues: { originalName: file.originalname, mimeType: file.mimetype, fileSize: file.size, url },
  });

  return media;
}

export async function updateMedia(id: number, actorId: string, data: { altText?: string | null }) {
  const media = await Media.findByPk(id);
  if (!media) {
    throw AppError.notFound('Media not found');
  }
  if (media.userId !== Number(actorId)) {
    throw AppError.forbidden();
  }
  if (data.altText !== undefined) {
    media.altText = data.altText;
  }
  await media.save();
  return media;
}

export async function deleteMedia(id: number, actorId: string) {
  const media = await Media.findByPk(id);
  if (!media) {
    throw AppError.notFound('Media not found');
  }

  const absolutePath = path.resolve(process.cwd(), env.uploadDir, media.fileName);
  await fs.unlink(absolutePath).catch(() => undefined);
  await media.destroy();
  await writeAuditLog({
    actorId: Number(actorId),
    action: 'MEDIA_DELETED',
    module: 'media',
    entityType: 'media',
    entityId: id,
    oldValues: { originalName: media.originalName, url: media.url },
  });
}

File: backend/app/services/media_service.py
import os
import uuid
from datetime import datetime

from app.config import MAX_UPLOAD_SIZE, UPLOAD_DIR
from app.core.audit import write_audit
from app.core.errors import AppError
from app.core.pagination import order_by_clause, parse_pagination
from app.models import Media

ALLOWED_EXTENSIONS = {
    '.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg',
    '.pdf', '.doc', '.docx', '.txt', '.mp4', '.webm',
}

PUBLIC_BASE = '/uploads'


def _ensure_upload_dir():
    directory = os.path.join(os.getcwd(), UPLOAD_DIR)
    if not os.path.exists(directory):
        os.makedirs(directory, exist_ok=True)
    return directory


def _extension(filename):
    return os.path.splitext(filename or '')[1].lower()


def list_media(db, filters):
    page, limit, offset, sort, order = parse_pagination(filters)

    q = db.query(Media).filter(Media.deletedAt.is_(None))
    if filters.get('search'):
        q = q.filter(Media.originalName.like('%{}%'.format(filters['search'])))
    if filters.get('mimeType'):
        q = q.filter(Media.mimeType.like('{}%'.format(filters['mimeType'])))

    total = q.count()
    media = q.order_by(order_by_clause(Media, sort, order)).offset(offset).limit(limit).all()

    return {
        'items': media,
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def get_media(db, media_id):
    media = db.query(Media).filter(Media.id == media_id, Media.deletedAt.is_(None)).first()
    if not media:
        raise AppError.not_found('Media not found')
    return media


def save_upload(db, actor_id, upload_file, alt_text=None):
    if upload_file is None:
        raise AppError.bad_request('No file uploaded')

    original_name = upload_file.filename or 'file'
    ext = _extension(original_name)
    if ext not in ALLOWED_EXTENSIONS:
        raise AppError.bad_request('File type not allowed: {}'.format(ext or 'unknown'))

    content = upload_file.file.read(MAX_UPLOAD_SIZE + 1)
    if len(content) > MAX_UPLOAD_SIZE:
        raise AppError.bad_request('File exceeds the maximum allowed size')

    directory = _ensure_upload_dir()
    filename = '{}{}'.format(uuid.uuid4().hex, ext)
    with open(os.path.join(directory, filename), 'wb') as f:
        f.write(content)

    url = '{}/{}'.format(PUBLIC_BASE, filename)

    media = Media(
        userId=actor_id,
        fileName=filename,
        originalName=original_name,
        mimeType=upload_file.content_type or 'application/octet-stream',
        fileSize=len(content),
        storagePath=os.path.join(UPLOAD_DIR, filename),
        url=url,
        altText=alt_text,
    )
    db.add(media)
    db.flush()

    write_audit(
        db, actor_id, 'MEDIA_UPLOADED', 'media', 'media', media.id,
        new={'originalName': original_name, 'mimeType': media.mimeType, 'fileSize': media.fileSize, 'url': url},
    )
    db.commit()
    return get_media(db, media.id)


def update_media(db, media_id, actor_id, data):
    media = db.query(Media).filter(Media.id == media_id, Media.deletedAt.is_(None)).first()
    if not media:
        raise AppError.not_found('Media not found')
    if media.userId != actor_id:
        raise AppError.forbidden()
    if 'altText' in data:
        media.altText = data['altText']
    db.commit()
    return get_media(db, media_id)


def delete_media(db, media_id, actor_id):
    media = db.query(Media).filter(Media.id == media_id, Media.deletedAt.is_(None)).first()
    if not media:
        raise AppError.not_found('Media not found')

    absolute_path = os.path.join(os.getcwd(), UPLOAD_DIR, media.fileName)
    try:
        if os.path.isfile(absolute_path):
            os.unlink(absolute_path)
    except OSError:
        pass

    media.deletedAt = datetime.utcnow()
    write_audit(db, actor_id, 'MEDIA_DELETED', 'media', 'media', media_id,
                old={'originalName': media.originalName, 'url': media.url})
    db.commit()